File: questions/q409_maximum_sum_subarray_with_negative_numbers/code_optimized_dp.cpp
class Solution {
    public:
        int maxSubArray(vector<int>& nums) {
            int curSubArrayMax = nums[0];
            int ans = curSubArrayMax;
            for (int i=1; i<nums.size(); i++) {
                curSubArrayMax = max(curSubArrayMax,0) + nums[i];
                ans = max(ans, curSubArrayMax);
            }
            return ans;
        }
    };

File: questions/q409_maximum_sum_subarray_with_negative_numbers/code_optimized_dp.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        s = 0
        maxSum = float('-inf')

        for num in nums :
            s += num
            maxSum = max(maxSum, s)

            if s < 0 :
                s = 0
        
        return maxSum
